stdlib/public/SwiftShims/RefCount.h

```diff
@@ -239,6 +282,48 @@ class StrongRefCount {
   }
 
   template <bool ClearPinnedFlag>
+  bool doDecrementShouldDeallocateNonAtomic() {
+    // If we're being asked to clear the pinned flag, we can assume
+    // it's already set.
+    constexpr uint32_t quantum =
+      (ClearPinnedFlag ? RC_ONE + RC_PINNED_FLAG : RC_ONE);
+    uint32_t val = __atomic_load_n(&refCount, __ATOMIC_RELAXED);
+    val -= quantum;
+    __atomic_store_n(&refCount, val, __ATOMIC_RELAXED);
+    uint32_t newval = refCount;
+
+    assert((!ClearPinnedFlag || !(newval & RC_PINNED_FLAG)) &&
+           "unpinning reference that was not pinned");
+    assert(newval + quantum >= RC_ONE &&
+           "releasing reference with a refcount of zero");
+
+    // If we didn't drop the reference count to zero, or if the
+    // deallocating flag is already set, we're done; don't start
+    // deallocation.  We can assume that the pinned flag isn't set
+    // unless the refcount is nonzero, and or'ing it in gives us a
+    // more efficient mask: the check just becomes "is newval nonzero".
+    if ((newval & (RC_COUNT_MASK | RC_PINNED_FLAG | RC_DEALLOCATING_FLAG))
+          != 0) {
+      // Refcount is not zero. We definitely do not need to deallocate.
+      return false;
+    }
+
+    // Refcount is now 0 and is not already deallocating.  Try to set
+    // the deallocating flag.  This must be atomic because it can race
+    // with weak retains.
+    //
+    // This also performs the before-deinit acquire barrier if we set the flag.
+    static_assert(RC_FLAGS_COUNT == 2,
+                  "fix decrementShouldDeallocate() if you add more flags");
+    uint32_t oldval = 0;
+    newval = RC_DEALLOCATING_FLAG;
+    if (refCount != oldval)
+      return false;
+    __atomic_store_n(&refCount, newval, __ATOMIC_RELAXED);
```

Comment:
Let's say we can prove that we called "make_unique" on such an object before, e.g. we have a COW data type like Array, Dictionary or Set and we prove that before the call of this operation there was a call that made sure that the buffers used by the COW objects internally are unique. Is it enough to make sure that something else provided the necessary barrier between threads?

Or should we better insert a barrier here to make it more general and to be on the safe side?